byway/src/main/webapp/recommendations.js

```diff
@@ -0,0 +1,257 @@
+      findRegions(result);
+      loadRecommendations();
+    } else {
+      alert("Could not calculate route due to: " + status);
+    }
+  });
+}
+
+/**
+ * Goes through steps of every leg along route to find the average location
+ * between a step's start and end location. Store these points
+ * in the regions Array.
+ * @param {DirectionsResult} directionResult contains directions
+ * for the route made.
+ */
+function findRegions(directionResult) {
+  const myRoute = directionResult.routes[0];
+  for(leg of myRoute.legs) {
+    for(step of leg.steps) {
+      const avgLat = (step.start_location.lat() + step.end_location.lat()) / 2;
+      const avgLng = (step.start_location.lng() + step.end_location.lng()) / 2;
+      const avgLoc = {lat: avgLat, lng: avgLng};
+      if(step.distance.value > MIN_DISTANCE_FOR_STEP_PATH) {
+        regions.push(avgLoc);
+      }
+    }
+  }
+}
+
+/**
+ * Set a timeout to delay the browser.
+ * @param {Number} delayMs number of milliseconds
+ * @return empty promise after delayMs milliseconds passed
+ */
+function delayPromise(delayMs) {
+  return new Promise(resolve => setTimeout(resolve, delayMs));
+}
+
+/**
+ * Go through a user's interests and search for places
+ * fitting those interests. Search around the regions
+ * previously found using textSearch. Prioritize finding
+ * past results loaded through sessionStorage to avoid calling
+ * textSearch repeatedly from the PlacesService.
+ */
+async function loadRecommendations() {
+  for(interest of interests) {
+    for(region of regions) {
+      const request = {
+        location: region,
+        radius: RADIUS_TO_SEARCH_AROUND,
+        query: interest
+      }
+      const recommendationsSaved = sessionStorage.getItem(JSON.stringify(request));
+      if(recommendationsSaved !== null) {
+        addRecommendations(request, JSON.parse(recommendationsSaved));
+      } else{
+        // Set an intermediate timeout between calls to findPlacesWithTextSearch
+        // to prevent hitting the query limit from the google maps API
+        await delayPromise(250);
+        const placesFound = await findPlacesWithTextSearch(request);
+        if(placesFound !== null) {
+          addRecommendations(request, placesFound);
+        }
+      }
+    }
+  }
+}
+
+/**
+ * Use the textSearch function from PlacesService to find the results
+ * fitting the request object passed in. Convert callback function into
+ * a chain of promises to return the result directly. If the service
+ * returns a status other than OK, reject the promise and alert the user.
+ * @param {TextSearchRequest} request object with location, radius and query fields.
+ * @return promise either with PlaceResult[] results or undefined if rejected.
+ */
+function findPlacesWithTextSearch(request) {
+  return new Promise((resolve, reject) => {
+    placesService.textSearch(request, (result, status) => {
+      if(status === "OK") {
+        resolve(result);
+      } else {
+        alertUser(status);
```

Comment:
You'll want to do the opposite in the case that it was successful. You'll probably want to reset the alerts in general every time you call this method. This way you don' show any errors that are no longer relevant. you could reset the visibility of the alerts before checking the status, and then only the newest, if any, errors wil be displayed

```javascript
function resetUserAlerts() {
    document.getElementById("message-container")style.visibility = 'hidden';
    document.getElementById("query-limit").style.visibility = 'visible';
    document.getElementById("incomplete-results").style.visibility = 'visible';
    document.getElementById("other-statuses").style.visibility = 'visible';
}
```

Reply:
This function is called inside a loop for multiple requests at a time. I want to alert the user if any request is not successfully filled, ie if one has zero results or it hits a query limit in the middle of the requests. If I add this code, the user may not get alerted if a status was not OK. Instead, I want a message to show up if at least one status other than OK was hit, and to list off all statuses hit if more than one.

I modified the code to simple show a general message, "showing limited results," and having it show the status returned from the PlacesService. I added a check to not redundantly show if a query limit was hit 10 times via sets.

Reply:
Thats a good catch with the loop. You wouldn't want to remove the alert if one of the iterations is successful. However, you may still want to hide the alerts if the user calculates the route multiple times. No reason to show the error if the second attempt succeeds. You could reset the alerts in the `loadRecommendations() ` functions instead, so that it would be outside this loop.

Reply:
Good point, forgot to account for that. Fixed